### PR TITLE
Added `PriorityQueue{Sink,Source}` extending `Queue{Sink,Source}`

### PR DESCRIPTION
**Pull request by neomaclin**
trying to continue the work #3733 by @lukecollier
@armanbilge

### CODE REVIEW

**Comment by durban** [comment]
It seems to me, that `tryTakeN` and `tryOfferN` has the same implementation in `QueueSource`/`QueueSink`. So why are we overriding?

**Comment by neomaclin** [comment]
> It seems to me, that `tryTakeN` and `tryOfferN` has the same implementation in `QueueSource`/`QueueSink`. So why are we overriding?

actually, come to think of it. they are actually identical in almost everyway. a simple extends without should just do., pushed

**Comment by neomaclin** [comment]
insteresting seeing the error in Mima check, maybe that is need of override?

**Comment by armanbilge** [comment]
> insteresting seeing the error in Mima check, maybe that is need of override?

yes, I think you can add an override call that delegates to the super so you do not need to repeat the implementation.

**Comment by neomaclin** [comment]
> > insteresting seeing the error in Mima check, maybe that is need of override?
> 
> yes, I think you can add an override call that delegates to the super so you do not need to repeat the implementation.

looks like Mima hates either way....(override with super call triggered the ReversedMissingMethodProblem)

**Comment by neomaclin** [comment]
why DispatcherSpec failed is a bit beyound me...

**Comment by neomaclin** [comment]
> It seems to me, that `tryTakeN` and `tryOfferN` has the same implementation in `QueueSource`/`QueueSink`. So why are we overriding?

I guess the attempts were telling us MiMa doesn't agree...

**Comment by armanbilge** [comment]
> looks like Mima hates either way....(override with super call triggered the ReversedMissingMethodProblem)

hmm. can you share the full MiMa error message? I don't understand why it's not happy.

**Comment by neomaclin** [comment]
> > looks like Mima hates either way....(override with super call triggered the ReversedMissingMethodProblem)
> 
> hmm. can you share the full MiMa error message? I don't understand why it's not happy.

ah. in order to fix the build, I have force pushed and squashed that commit (not sure how to find the actual error log in that case). but bascially MiMa is reported the "ReversedMissingMethodProblem" along with the message "super.tryOfferN is present only in the current version" if I have only

`override def tryOfferN(list: List[A])(implicit F: Monad[F]): F[List[A]] = super.tryOfferN(list)`

**Comment by neomaclin** [comment]
> > looks like Mima hates either way....(override with super call triggered the ReversedMissingMethodProblem)
> 
> hmm. can you share the full MiMa error message? I don't understand why it's not happy.

or, I could break the build again to generate the error.

**Comment by neomaclin** [comment]
> > looks like Mima hates either way....(override with super call triggered the ReversedMissingMethodProblem)
> 
> hmm. can you share the full MiMa error message? I don't understand why it's not happy.

here they are

```log
[error] cats-effect-std: Failed binary compatibility check against org.typelevel:cats-effect-std_2.13:3.3.11 (e:info.apiURL=https://typelevel.org/cats-effect/api/3.x/, e:info.versionScheme=early-semver)! Found 2 potential problems (filtered 18)
[error]  * abstract synthetic method cats$effect$std$PQueueSink$$super$tryOfferN(scala.collection.immutable.List,cats.Monad)java.lang.Object in interface cats.effect.std.PQueueSink is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("cats.effect.std.PQueueSink.cats$effect$std$PQueueSink$$super$tryOfferN")
[error]  * abstract synthetic method cats$effect$std$PQueueSource$$super$tryTakeN(scala.Option,cats.Monad)java.lang.Object in interface cats.effect.std.PQueueSource is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("cats.effect.std.PQueueSource.cats$effect$std$PQueueSource$$super$tryTakeN")
```

**Comment by neomaclin** [comment]
[fb356b7](https://github.com/typelevel/cats-effect/pull/3930/commits/fb356b75ea75b13dcd0e9f775989ea298f0fdd96) with the implementaion on the overrided method was ok with MiMa (and other build checks). let me know if I should go back to that for succcess build of the PR. oh. and since I don't have the write access, I wouldn't be able to merge anything even if the build is/was ok.

**Comment by armanbilge** [comment]
We should avoid duplicating the implementation. Since the `super` trick is not possible, we can extract the implementation to a method on the companion object, and then both traits can delegate to that.

**Comment by neomaclin** [comment]
some how the DispatcherSpec comes back

```
[error]     x raise an error on leaked runner
[error]      Expected: java.lang.IllegalStateException. Got nothing (DispatcherSpec.scala:269)
```

**Comment by samspills** [comment]
> some how the DispatcherSpec comes back
> 
> ```
> [error]     x raise an error on leaked runner
> [error]      Expected: java.lang.IllegalStateException. Got nothing (DispatcherSpec.scala:269)
> ```

@neomaclin that's unrelated to your PR, don't worry :) It's a known flake: https://github.com/typelevel/cats-effect/issues/3776

**Comment by neomaclin** [comment]
if this is good to go in, please go ahead, ( I can't do it myself..)

File: std/shared/src/main/scala/cats/effect/std/PQueue.scala [hunk 1]
```diff
@@ -302,54 +251,26 @@ object PQueueSource {
...
-   * method that recursively runs `tryOffer` and does not offer any additional performance
-   * benefits.
+   * Attempts to enqueue the given elements at the back of the queue without semantically
+   * blocking. If an item in the list cannot be enqueued, the remaining elements will be
+   * returned. This is a convenience method that recursively runs `tryOffer` and does not offer
+   * any additional performance benefits.
    *
    * @param list
-   *   the elements to be put in the PQueue
+   *   the elements to be put at the back of the queue
```
Comment (durban) [inline std/shared/src/main/scala/cats/effect/std/PQueue.scala]: "at the back of the queue" is not exactly correct here, as this is a priority queue.

Comment (neomaclin) [reply std/shared/src/main/scala/cats/effect/std/PQueue.scala]: > "at the back of the queue" is not exactly correct here, as this is a priority queue.

reverting the related docs for tryOfferN in PQueueSink

**Review by durban — APPROVED** [review]

**Comment by djspiewak** [comment]
Thank you for taking this on!

**Pull request merged**